src/pytest_ibutsu/modeling.py

```diff
@@ -0,0 +1,289 @@
+import os
+import time
+import uuid
+from datetime import datetime
+from typing import ClassVar
+from typing import Dict
+from typing import List
+from typing import Optional
+from typing import Union
+
+import pytest
+from attrs import asdict
+from attrs import define
+from attrs import field
+
+
+def safe_string(o):
+    """This will make string out of ANYTHING without having to worry about the stupid Unicode errors
+
+    This function tries to make str/unicode out of ``o`` unless it already is one of those and then
+    it processes it so in the end there is a harmless ascii string.
+
+    Args:
+        o: Anything.
+    """
+    if not isinstance(o, str):
+        o = str(o)
+    if isinstance(o, bytes):
+        o = o.decode("utf-8", "ignore")
+    o = o.encode("ascii", "xmlcharrefreplace").decode("ascii")
+    return o
+
+
+@define
+class Summary:
+    failures: int = 0
+    errors: int = 0
+    xfailures: int = 0
+    xpasses: int = 0
+    tests: int = 0
+    collected: int = 0
+    not_run: int = 0
+
+    def increment(self, test_result: "TestResult") -> None:
+        attr = {
+            "failed": "failures",
+            "error": "errors",
+            "skipped": "skips",
+            "xfailed": "xfailures",
+            "xpassed": "xpasses",
+        }.get(test_result.result)
+        if attr:
+            inc = getattr(self, attr)
+            setattr(self, attr, inc + 1)
+        self.tests += 1
+
+
+@define
+class TestRun:
+    component: Optional[str] = None
+    env: Optional[str] = None
+    id: str = field(factory=lambda: str(uuid.uuid4()))
```

Comment:
Not finding any instantiations of this class in the PR, would `id` be passed ever? (`init=False` kwarg for field)